goldenmaster/apigear/mqtt/mqttclient.cpp

```diff
@@ -0,0 +1,211 @@
+    auto subscribedTopic = subscription->topic();
+    if (!m_subscriptions.contains(subscribedTopic))
+    {
+        connect(subscription, &QMqttSubscription::messageReceived,
+                [this](const QMqttMessage& message)
+                {
+                    auto subscription = m_subscriptions.begin();
+                    for (subscription = m_subscriptions.begin(); subscription!= m_subscriptions.end(); subscription++)
+                    {
+                        if (subscription.key().match(message.topic()))
+                            break;
+                    }
+
+                    while (subscription != m_subscriptions.end() && subscription.key().match(message.topic()))
+                    {
+                        if (subscription.value().second)
+                        {
+                            auto arguments = PayloadConverter::fromPayload(message.payload());
+                            subscription.value().second(arguments);
+                        }
+                        subscription++;
+                    }
+                });
+
+    }
+    m_subscriptions.insert(subscription->topic(), std::make_pair(id, callback));
+}
+
+void Client::onSubscribeForInvokeResponse(quint64 id, const QString &topic)
+{
+    auto subscription = m_client.subscribe(topic, QoS);
+    auto topicFilter = subscription->topic();
+    if (!m_invokeReplySubscriptions.contains(topicFilter))
+    {
+        connect(subscription, &QMqttSubscription::messageReceived, this, &Client::handleInvokeResp);
+    }
+    m_invokeReplySubscriptions.insert(topicFilter, id);
+}
+
+void Client::onUnsubscribedTopic(quint64 subscriptionId)
+{
+
+    auto found = std::find_if(m_subscriptions.begin(),
+        m_subscriptions.end(), 
+        [subscriptionId](auto& element){ return element.first == subscriptionId;});
+    if (found != m_subscriptions.end())
+    {
+        auto topic = found.key();
+        m_subscriptions.erase(found);
+        auto values = m_subscriptions.values(topic);
+        if (values.empty())
+        {
+            m_client.unsubscribe(topic);
+        }
+    }
+    else
+    {
+        auto foundInvokeSubscription = std::find_if(m_invokeReplySubscriptions.begin(),
+            m_invokeReplySubscriptions.end(),
+            [subscriptionId](auto& element){ return element == subscriptionId;});
+
+        if (foundInvokeSubscription != m_invokeReplySubscriptions.end())
+        {
+            auto topic = foundInvokeSubscription.key();
+            m_invokeReplySubscriptions.erase(foundInvokeSubscription);
+            auto values = m_invokeReplySubscriptions.values(topic);
+            if (values.empty())
+            {
+                m_client.unsubscribe(topic);
+            }
+        }
+    }
+}
+
+bool Client::isReady() const
+{
+    return m_client.state() == QMqttClient::ClientState::Connected;
+}
+
+const QString& Client::clientId() const
+{
+    static auto clientID =  m_client.clientId();
```

Comment:
so I assumed it is always one give by user,  but the documentation says:
`Each client needs to have a unique ID to be able to connect to an MQTT broker. If no client ID is specified by the user, one will be generated automatically when a connection is established.`
so in general it can be different 🤔 assuming we allow not giving a name this mean not only static here may be wrong, but also the way I subscribe for invoke resp 🤔 and in general reconnection scenarios are missing 😞

Reply:
changed